File: frontend/src/types/workout.ts
export type Exercise = {
  id: string;
  name_ru: string;
  muscle_group: string;
  secondary_muscle_groups?: string[];
  equipment: string | null;
  description: string | null;
  technique: string | null;
  common_mistakes: string | null;
  difficulty: number;
  video_url: string | null;
  animation_url: string | null;
  thumbnail_url: string | null;
  media_duration_sec: number | null;
  media_source: "youtube" | "external" | "none" | string;
  tags: string[];
  limitations?: string[];
  weight_rule?: "total" | "per_hand" | "per_side" | "none";
};

export type WorkoutSet = {
  id: string;
  workout_id: string;
  exercise_id: string;
  set_number: number;
  reps: number | null;
  weight: number | null;
  weight_mode?: "total" | "per_hand" | null;
  is_completed: boolean;
  rest_time_sec: number | null;
  duration_sec?: number | null;
  note?: string | null;
  machine_params?: Record<string, string | number> | null;
  /** Local-only values retained while an exercise is temporarily replaced. */
  replacement_original_weight?: number | null;
  replacement_original_machine_params?: Record<string, string | number> | null;
};

export type WorkoutPlanExercise = {
  exercise_id: string;
  order: number;
  target_sets: number;
  target_reps?: string | null;
  rest_sec?: number | null;
  name_ru?: string | null;
  suggested_weight?: number | null;
  /** Set when user replaces an exercise during a session; used to restore default. */
  original_exercise_id?: string | null;
  weight_mode?: "total" | "per_hand" | null;
  note?: string | null;
};

export type WorkoutPlan = {
  title?: string | null;
  workout_type?: string | null;
  day_index?: number | null;
  /** light | medium | heavy — 3-week cycle */
  week_phase?: "light" | "medium" | "heavy" | string | null;
  week_in_cycle?: number | null;
  week_label?: string | null;
  week_rir?: string | null;
  /** Planned progression phase before an optional day-level adjustment. */
  base_week_phase?: "light" | "medium" | "heavy" | string | null;
  load_adjustment?: string | null;
  load_adjustment_label?: string | null;
  location?: string | null;
  equipment?: string[];
  limitations?: string[];
  exercises: WorkoutPlanExercise[];
};

export type Workout = {
  id: string;
  user_id: string;
  program_id: string | null;
  scheduled_date: string;
  status: "planned" | "completed" | "skipped" | string;
  ai_notes: string | null;
  rpe: number | null;
  started_at: string | null;
  completed_at: string | null;
  title?: string | null;
  workout_type?: string | null;
  plan?: WorkoutPlan | Record<string, unknown> | null;
  duration_sec?: number | null;
  sets: WorkoutSet[];
};

export type LocalSetDraft = {
  exerciseId: string;
  setNumber: number;
  reps: string;
  weight: string;
  weightMode?: "total" | "per_hand" | null;
  isCompleted: boolean;
  restTimeSec: number;
  /** Duration seconds for timed / cardio sets (optional). */
  durationSec?: number | null;
  note?: string | null;
  /** Free-form machine params: speed, incline, resistance… */
  machineParams?: Record<string, string | number> | null;
  /** Values preserved only so a bulk replacement can be fully reverted. */
  replacementOriginalWeight?: string;
  replacementOriginalMachineParams?: Record<string, string | number> | null;
};

export type Program = {
  id: string;
  name: string;
  description: string | null;
  target_level: string | null;
  duration_weeks: number | null;
  structure: Record<string, unknown>;
  workout_type: string;
  level: string | null;
  is_template: boolean;
  publication_status?: "draft" | "published" | "archived";
  program_key?: string;
  version?: number;
  is_current?: boolean;
  published_at?: string | null;
};

File: backend/app/schemas/workout.py
"""Workout and set request/response schemas."""

from __future__ import annotations

import uuid
from datetime import date, datetime
from decimal import Decimal

from pydantic import BaseModel, ConfigDict, Field


class WorkoutPlanExercise(BaseModel):
    exercise_id: uuid.UUID
    order: int = Field(..., ge=1)
    target_sets: int = Field(default=3, ge=1, le=20)
    target_reps: str | None = "8-12"
    rest_sec: int = Field(default=60, ge=0, le=600)
    name_ru: str | None = None
    suggested_weight: Decimal | None = Field(default=None, ge=0)
    original_exercise_id: uuid.UUID | None = None
    weight_mode: str | None = Field(default=None, pattern=r"^(total|per_hand)$")
    note: str | None = Field(default=None, max_length=500)


class WorkoutPlan(BaseModel):
    title: str | None = None
    workout_type: str | None = None
    day_index: int | None = None
    week_phase: str | None = None  # light | medium | heavy
    week_in_cycle: int | None = None
    week_label: str | None = None
    week_rir: str | None = None
    location: str | None = None
    equipment: list[str] = Field(default_factory=list)
    limitations: list[str] = Field(default_factory=list)
    exercises: list[WorkoutPlanExercise] = Field(default_factory=list)


class WorkoutCreate(BaseModel):
    scheduled_date: date
    client_workout_id: uuid.UUID | None = None
    program_id: uuid.UUID | None = None
    day_index: int | None = Field(default=None, ge=1)
    week_phase: str | None = Field(default=None, pattern=r"^(light|medium|heavy)$")
    title: str | None = None
    workout_type: str | None = None
    exercise_ids: list[uuid.UUID] = Field(
        default_factory=list,
        description="Optional exercises to pre-create empty set slots for",
    )
    plan: WorkoutPlan | None = None
    sets_per_exercise: int = Field(default=3, ge=1, le=10)


class PlannedWorkoutReplacement(BaseModel):
    from_exercise_id: uuid.UUID
    to_exercise_id: uuid.UUID


class PlannedWorkoutPlanRequest(BaseModel):
    program_id: uuid.UUID
    scheduled_date: date
    day_index: int = Field(..., ge=1)
    week_phase: str | None = Field(default=None, pattern=r"^(light|medium|heavy)$")
    replacements: list[PlannedWorkoutReplacement] = Field(default_factory=list, max_length=50)


class WorkoutCompleteRequest(BaseModel):
    rpe: int | None = Field(default=None, ge=1, le=10)
    ai_notes: str | None = None


class WorkoutUpdateRequest(BaseModel):
    """Editable summary fields for an existing workout."""

    rpe: int | None = Field(default=None, ge=1, le=10)
    ai_notes: str | None = Field(default=None, max_length=5000)


class WorkoutSetCreate(BaseModel):
    exercise_id: uuid.UUID
    set_number: int = Field(..., gt=0)
    reps: int | None = Field(default=None, ge=0, le=100_000)
    weight: Decimal | None = Field(default=None, ge=0, le=10_000)
    weight_mode: str | None = Field(default=None, pattern=r"^(total|per_hand)$")
    rest_time_sec: int | None = Field(default=None, ge=0, le=3600)
    duration_sec: int | None = Field(default=None, ge=0, le=86400)
    note: str | None = Field(default=None, max_length=1000)
    machine_params: dict[str, str | int | float] | None = None
    is_completed: bool = False


class WorkoutSetResponse(BaseModel):
    model_config = ConfigDict(from_attributes=True)

    id: uuid.UUID
    workout_id: uuid.UUID
    exercise_id: uuid.UUID
    set_number: int
    reps: int | None = None
    weight: Decimal | None = None
    weight_mode: str | None = None
    is_completed: bool
    rest_time_sec: int | None = None
    duration_sec: int | None = None
    note: str | None = None
    machine_params: dict[str, str | int | float] | None = None
    created_at: datetime
    updated_at: datetime


class WorkoutResponse(BaseModel):
    model_config = ConfigDict(from_attributes=True)

    id: uuid.UUID
    user_id: uuid.UUID
    program_id: uuid.UUID | None = None
    scheduled_date: date
    status: str
    ai_notes: str | None = None
    rpe: int | None = None
    started_at: datetime | None = None
    completed_at: datetime | None = None
    title: str | None = None
    workout_type: str | None = None
    plan: dict = Field(default_factory=dict)
    duration_sec: int | None = None
    created_at: datetime
    updated_at: datetime
    sets: list[WorkoutSetResponse] = Field(default_factory=list)


class WorkoutHistoryResponse(BaseModel):
    items: list[WorkoutResponse]
    total: int
